### PR TITLE
Fix pyfunc bugs for release

### PR DESCRIPTION
**Pull request by deadlycoconuts**
### Context

This PR aims to include a couple of minor bug fixes and refactoring before the upcoming minor release of Turing:
- Renaming of the `features` keyword argument of the abstract `ensemble` method in `EnsemblerBase` of the SDK to `input` for less ambiguity when implementing Pyfunc ensemblers
- Transformation of the `list` of route responses in real-time ensembler requests to a `dict` of route responses indexed by route names in the SDK `PyFunc` interface, so as to improve usability and consistency of the `predictions` keyword argument of the aforementioned `ensemble` method
- Addition of a name search query to the Turing API endpoint for listing registered ensemblers, so as to resolve the non-functioning ensembler search bar in the Turing UI
- Fixing of e2e batch prediction tests due to the non-deterministic behaviour of the experiment engine
- (Temporary) downgrading of support for the Turing SDK, ensembler job engine and ensembler service engine to **_only_** Python `3.7.*`. This is to align Turing with Merlin's support for Python of that particular minor version.

### CODE REVIEW

File: api/api/specs/ensemblers.yaml
Comment (deadlycoconuts) [inline api/api/specs/ensemblers.yaml:38]: Added this to the API endpoint so the ensembler search bar should now work.

File: engines/pyfunc-ensembler-service/tests/conftest.py
Comment (deadlycoconuts) [inline engines/pyfunc-ensembler-service/tests/conftest.py:32]: Removed this unnecessary helper method since the `list` of route responses -> `dict` of route responses transformation is now all done behind the scenes by the `PyFunc` interface.

File: sdk/turing/version.py
Comment (deadlycoconuts) [inline sdk/turing/version.py:1]: I think I forgot to update this for the release of Turing `0.2.0` but it's now going to updated as usual.

File: infra/e2e/turing.values.yaml
Comment (deadlycoconuts) [inline infra/e2e/turing.values.yaml:41]: @romanwozniak , I didn't manage to pick these changes up in your previous PR, but were you intending to demonstrate some treatment assignment behaviour when you introduced these configs and swapped `nop` for `proprietary` here? https://github.com/gojek/turing/blob/d12014c897a815d3526a41d390dc360e53a3a1a7/api/e2e/test/testdata/create_router_nop_logger_proprietary_exp.json.tmpl#L15
I think it's causing some odd undeterministic behaviour in the e2e tests in the CI pipeline like here:
https://github.com/gojek/turing/runs/5673171262?check_suite_focus=true#step:18:74

File: api/turing/service/ensembler_service.go
Comment (krithika369) [inline api/turing/service/ensembler_service.go:75]: `ilike` may be better, for case-insensitive search.

File: sdk/turing/version.py
Comment (krithika369) [reply sdk/turing/version.py:1]: How often do we manually update it? And what would be the benefit of doing so? I wonder if this can just be left at `0.0.0` always, since the `make version` command overwrites this file with the correct version during build anyway.

File: api/api/specs/ensemblers.yaml
Comment (krithika369) [reply api/api/specs/ensemblers.yaml:38]: Would be good to include a description for this field as it's not immediately clear from the name what we are searching. For example, something like:`Searches ensembler name for a partial match of the search text`

**Review by krithika369 — APPROVED** [review]
Left some small comments. LGTM, pending the discussion about the e2e test case. Thank you!

File: sdk/turing/version.py
Comment (deadlycoconuts) [reply sdk/turing/version.py:1]: > How often do we manually update it?

I think judging from... how the original string was `0.1.0`, and that's our first release... we've probably never updated it at all 😅

>And what would be the benefit of doing so? I wonder if this can just be left at 0.0.0 always, since the make version command overwrites this file with the correct version during build anyway.

That is true. Maybe I'll just roll it back to `0.0.0` like you suggested, and leave a comment to say that's no need to update that field manually.

File: infra/e2e/turing.values.yaml
Comment (romanwozniak) [reply infra/e2e/turing.values.yaml:41]: @deadlycoconuts https://github.com/gojek/turing/pull/189

Comment (romanwozniak) [reply infra/e2e/turing.values.yaml:41]: Don't forget to revert this before merging @deadlycoconuts

File: api/api/openapi.bundle.yaml
Comment (krithika369) [inline api/api/openapi.bundle.yaml:180]: Isn't it just ensembling job name? 😄 The job name is generated from the name of the ensembler, with some suffix. But ultimately, the search is carried out on the ensembling job's name.

Comment (deadlycoconuts) [reply api/api/openapi.bundle.yaml:180]: Oh that's true too xD Haha alright I'll simplify the name back to what it should plainly be 😅

File: engines/pyfunc-ensembler-job/environment.yaml
Comment (krithika369) [inline engines/pyfunc-ensembler-job/environment.yaml:3]: I think we need to update these for the engines too: https://github.com/gojek/turing/blob/main/engines/pyfunc-ensembler-job/setup.py#L25

**Comment by deadlycoconuts** [comment]
@romanwozniak https://github.com/gojek/turing/runs/5715592714?check_suite_focus=true#step:18:78 🥲  I'll go look at it again and see what went wrong.

**Comment by romanwozniak** [comment]
> @romanwozniak https://github.com/gojek/turing/runs/5715592714?check_suite_focus=true#step:18:78 🥲 I'll go look at it again and see what went wrong.

I have only two assumptions:
 * `"hash/fnv"` hashing is non-deterministic (I doubt that)
 * There is something wrong with the `/batch_predict` endpoint – i.e. response from one prediction request in the batch is being assigned to the wrong request before marshaling
 This is the raw response, so the issue has nothing to do with the JSON comparison

```
01_create_router_test.go:84: Response Payload:
        [{"code":200,"data":{"request":{"client":{"id":4}},"response":{"route_responses":[{"route":"control","data":{"version":"control"},"is_default":true}],"experiment":{"configuration":{"foo":"bar"}}}}},{"code":200,"data":{"request":{"client":{"id":7}},"response":{"route_responses":[{"route":"control","data":{"version":"control"},"is_default":true}],"experiment":{"configuration":{"foo":"bar"}}}}}]
```

You probably can merge this MR and keep the fix to the E2E test out of scope for now

**Comment by romanwozniak** [comment]
@deadlycoconuts actually, it's neither nor. I can reproduce the issue by running the unit test ( which I added recently) multiple times. However, the hashing is consistent. It could be related to the floating-point precision or something similar. Let  me try to figure it out

**Comment by deadlycoconuts** [comment]
Thanks a lot for your help for the time being @romanwozniak !

**Comment by romanwozniak** [comment]
> Thanks a lot for your help for the time being @romanwozniak !

TIL: `range` operation over `map` is not deterministic

**Comment by romanwozniak** [comment]
> Thanks a lot for your help for the time being @romanwozniak !

@deadlycoconuts It's fixed now. Sorry, for hijacking your PR. The diff is here: 
https://github.com/gojek/turing/pull/188/files/481c6933929dbea18854ad9ebab07a1b3ba6e612..808818c4285983ef243ab6612545b4fff91e51a0

**Comment by deadlycoconuts** [comment]
Haha no worries, thanks a lot for digging into and solving the mystery of this really elusive bug! 😂

**Comment by deadlycoconuts** [comment]
Okay I'll be merging this if there aren't anything outstanding! Thanks a lot everyone for your input! 🚀

**Pull request merged**
